vsintegration/src/FSharp.Editor/LanguageService/MetadataAsSource.fs

```diff
@@ -0,0 +1,159 @@
+
+        let assemblyInfoDocument = 
+            DocumentInfo.Create(
+                assemblyInfoDocumentId,
+                assemblyInfoFileName,
+                loader = TextLoader.From(assemblyInfoSourceTextContainer, VersionStamp.Default))
+
+        let generatedDocumentId = DocumentId.CreateNewId(projectId)
+        let documentInfo = 
+            DocumentInfo.Create(
+                generatedDocumentId,
+                Path.GetFileName(temporaryFilePath),
+                filePath = temporaryFilePath,
+                loader = FileTextLoader(temporaryFilePath, Encoding.UTF8))
+
+        let projectInfo = 
+            ProjectInfo.Create(
+                projectId,
+                VersionStamp.Default,
+                name = asmIdentity.Name,
+                assemblyName = asmIdentity.Name,
+                language = LanguageNames.CSharp,
+                compilationOptions = compilationOptions,
+                parseOptions = parseOptions,
+                documents = [|assemblyInfoDocument;documentInfo|],
+                metadataReferences = metadataReferences)
+
+        (projectInfo, documentInfo)
+
+    let DecompileCSharp (symbolFullTypeName: string, assemblyLocation: string) =
+        let logger = new StringBuilder();
+
+        // Initialize a decompiler with default settings.
+        let decompiler = CSharpDecompiler(assemblyLocation, DecompilerSettings())
+        // Escape invalid identifiers to prevent Roslyn from failing to parse the generated code.
+        // (This happens for example, when there is compiler-generated code that is not yet recognized/transformed by the decompiler.)
+        decompiler.AstTransforms.Add(new EscapeInvalidIdentifiers())
+
+        let fullTypeName = FullTypeName(symbolFullTypeName)
+
+        // Try to decompile; if an exception is thrown the caller will handle it
+        let mutable text = decompiler.DecompileTypeAsString(fullTypeName)
+
+        text <- text + "#if false // " + Environment.NewLine
+        text <- text + logger.ToString()
+        text <- text + "#endif" + Environment.NewLine
+
+        SourceText.From(text)
+
+    let ShowDocument (filePath, name, serviceProvider: IServiceProvider) =
+        let vsRunningDocumentTable4 = serviceProvider.GetService<SVsRunningDocumentTable, IVsRunningDocumentTable4>()
+        let fileAlreadyOpen = vsRunningDocumentTable4.IsMonikerValid(filePath)
+
+        let openDocumentService = serviceProvider.GetService<SVsUIShellOpenDocument, IVsUIShellOpenDocument>()
+        let mutable localServiceProvider = Unchecked.defaultof<_>
+        let mutable hierarchy = Unchecked.defaultof<_>
+        let mutable itemId = Unchecked.defaultof<_>
+        let mutable windowFrame = Unchecked.defaultof<_>
+        openDocumentService.OpenDocumentViaProject(filePath, ref VSConstants.LOGVIEWID.TextView_guid, &localServiceProvider, &hierarchy, &itemId, &windowFrame) |> ignore
```

Comment:
Can you use output tuples?

        let _, localServiceProvider, hierarchy, itemId, windowFrame = openDocumentService.OpenDocumentViaProject(filePath, ref VSConstants.LOGVIEWID.TextView_guid)